lib/galaxy/managers/users.py

```diff
@@ -236,6 +247,150 @@ def tags_used(self, user, tag_models=None):
         tags = [((name + ':' + val) if val else name) for name, val in tags]
         return sorted(tags)
 
+    def change_password(self, trans, password=None, confirm=None, token=None, id=None, current=None):
+        """
+        Allows to change a user password with a token.
+        """
+        if not token and not id:
+            return None, "Please provide a token or a user and password."
+        if token:
+            token_result = trans.sa_session.query(trans.app.model.PasswordResetToken).get(token)
+            if not token_result or not token_result.expiration_time > datetime.utcnow():
+                return None, "Invalid or expired password reset token, please request a new one."
+            user = token_result.user
+            message = self.__set_password(trans, user, password, confirm)
+            if message:
+                return None, message
+            token_result.expiration_time = datetime.utcnow()
+            trans.sa_session.add(token_result)
+            return user, "Password has been changed. Token has been invalidated."
+        else:
+            user = self.by_id(self.app.security.decode_id(id))
+            if user:
+                message = self.app.auth_manager.check_change_password(user, current)
+                if message:
+                    return None, message
+                message = self.__set_password(trans, user, password, confirm)
+                if message:
+                    return None, message
+                return user, "Password has been changed."
+            else:
+                return user, "User not found."
+
+    def __set_password(self, trans, user, password, confirm):
+        if not password:
+            return "Please provide a new password."
+        if user:
+            # Validate the new password
+            message = validate_password(trans, password, confirm)
+            if message:
+                return message
+            else:
+                # Save new password
+                user.set_password_cleartext(password)
+                # Invalidate all other sessions
+                if trans.galaxy_session:
+                    for other_galaxy_session in trans.sa_session.query(trans.app.model.GalaxySession) \
+                                                     .filter(and_(trans.app.model.GalaxySession.table.c.user_id == user.id,
+                                                                  trans.app.model.GalaxySession.table.c.is_valid == true(),
+                                                                  trans.app.model.GalaxySession.table.c.id != trans.galaxy_session.id)):
+                        other_galaxy_session.is_valid = False
+                        trans.sa_session.add(other_galaxy_session)
```

Comment:
Since you're updating an existing record, you may not need to `add` it to the database, though it needs to be tested.